File: java dsa codes/deletionlinkedlist.java
public class deletionlinkedlist {

    static class Node{
        int data;
        Node next;
    }

    public static void linkedlisttraversl(Node head){
        while(head!=null){
            System.out.println(head.data);
            head=head.next;
        }
    }

    public static Node deleteatfirst(Node head){
        if(head==null){
            return null;
        }

        Node p =head;
        head=head.next;
        return head;

    }

    public static Node deleteatindex(Node head,int index){
        Node p=head;
        Node q=p.next;
        int i=0;
        while(i!=index-1){
            p=p.next;
            i++;
        }
        p.next=q.next;
        return head;
    }

    public static Node deleteatend(Node head){
        Node p=head;
        Node q=p.next;
        if(head==null || head.next==null){
            return null;
        }

        while(p.next.next==null){
            p=p.next;
            q=q.next;
        }
        p.next=null;
        return head;

    }

    public static Node deleteatvalue(Node head,int value){
        if(head==null ){
            return null;
        }
        if(head.data==value){
            return head.next;
        }
        Node p=head;
        Node q=p.next;
        while(q.data!=value && q!=null){
            p=q;
            q=q.next;
        }
        if(q.data==value){
            p.next=q.next;
            return head;
        }
        return head;
    }

    public static void main(String [] args){
        Node head=new Node();
        head.data=11;
        head.next=null;

        Node second=new Node();
        second.data=22;
        second.next=null;

        Node third=new Node();
        third.data=45;
        third.next=null;

        head.next=second;
        second.next=third;

        head=deleteatvalue(head,45);
        linkedlisttraversl(head);
    }
}

File: c dsa codes/deletionlinkedlist.c
#include<stdio.h>
#include<stdlib.h>

struct node{
    int data;
    struct node* Next;
};

int deleteatfirst(struct node* head){
    struct node * ptr=head;
    head=head->Next;
    free(ptr);
    return head;
}

int deleteatindex(struct node* head,int index){
    struct node* p=head;
    struct node* q=p->Next;

    int i=0;
    while(i!=index-1){
        p=p->Next;
        i++;
    }
    p->Next=q->Next;
    free(q);
    return head;
}

struct node* deleteatend(struct node* head) {

    if (head == NULL) {
        return NULL;
    }

    if (head->Next == NULL) {
        free(head);
        return NULL;
    }

    struct node* p = head;

    while (p->Next->Next != NULL) {
        p = p->Next;
    }

    struct node* q = p->Next;
    p->Next = NULL;
    free(q);

    return head;
}


int deleteatvalue(struct node* head,int value){
    struct node *p=head;
    struct node* q=p->Next;

    while(q->data!=value && q->Next!=NULL){
        p=p->Next;
    }
    
    if(q->data==value){
        p->Next=q->Next;
        free(q);
        return head;
    }
}

void linkedlistTraversal(struct node* ptr){

    

    while (ptr != NULL)
    {
        printf("Element: %d\n", ptr->data);
        ptr = ptr->Next;
    }
}

int main(){
// Linked list making
    struct node* head;
    struct node* second;
    struct node* third;
    struct node* fourth;

    head=(struct node*)malloc(sizeof(struct node));
    second=(struct node*)malloc(sizeof(struct node));
    third=(struct node*)malloc(sizeof(struct node));
    fourth=(struct node*)malloc(sizeof(struct node));

    head->data=32;
    head->Next=second;

    second->data=22;
    second->Next=third;

    third->data=11;
    third->Next=fourth;

    fourth->data=44;
    fourth->Next=NULL;

    head=deleteatindex(head ,2);

    printf("After deletion at end:\n");

    linkedlistTraversal(head);

}
